draft-ietf-add-ddr.md

```diff
@@ -281,30 +299,58 @@ Resolver owners that support authenticated discovery will need to list valid
 referring IP addresses in their TLS certificates. This may pose challenges for
 resolvers with a large number of referring IP addresses.
 
+## Legacy Partial Forwarders
+
+A "partial forwarder" is a DNS device that forwards some queries to an upstream
+resolver, and answers some queries directly.  This includes both filtering and
+split-horizon DNS behaviors.  As noted in {{opportunistic}}, such devices SHOULD
+intercept queries to `_dns.resolver.arpa` to ensure that the upstream resolver
+does not trigger an opportunistic upgrade. However, some networks contain legacy
+partial forwarders whose behavior cannot be updated.  For compatibility with
+legacy partial forwarders, clients MAY impose additional conditions on
+opportunistic discovery, such as requiring that the Encrypted Resolver is on the
+same private IP address as the Unencrypted Resolver.
+
+## Limited compatibility with getaddrinfo for IP-based discovery
+
+Clients that access DNS through the `getaddrinfo` API can only issue A and AAAA
+queries, and can only receive these records and CNAMEs. Such clients MAY
+attempt IP-based discovery ({{bootstrapping}}) by issuing an A or AAAA query
+for `_dns.resolver.arpa`. If a CNAME record is returned, it indicates the
+DNS server hostname (plus a `_dns` prefix). (The A or AAAA query is not
+generally expected to succeed, and is used only to elicit the CNAME.) If the
+client already knows how to connect to this server, or is able to determine
+how to connect through some other mechanism, it MAY do so.
+
 # Security Considerations
 
-Since client can receive DNS SVCB answers over unencrypted DNS, on-path
-attackers can prevent successful discovery by dropping SVCB packets. Clients
-should be aware that it might not be possible to distinguish between resolvers
-that do not have any Designated Resolver and such an active attack.
+Since clients can receive DNS SVCB answers over unencrypted DNS, in-path
+attackers can prevent successful discovery by dropping SVCB packets. When using
+IP-based discovery ({{bootstrapping}}), it is not possible to distinguish between resolvers
+that do not have a Designated Resolver and such an active attack.
+Clients that wish to defend against these attacks MUST use name-based discovery
+({{encrypted}}) with local DNSSEC validation, or otherwise have an authenticated
```

Comment:
Does this mean that a DNSSEC signed record at _dns.$HOSTNAME is a pin in some circumstances?  That is, for the duration of its TTL, if the server is not available, clients might not be able to access DNS service.

Reply:
As [currently specified](https://datatracker.ietf.org/doc/html/draft-schwartz-svcb-dns-02#section-9.2), it's a pin even if not DNSSEC signed.  The theory here is that, even without DNSSEC, you could easily have an adversary who can block your DoT connection but not your bootstrap queries.  You wouldn't want that adversary to cause an automatic fallback to plaintext DNS.